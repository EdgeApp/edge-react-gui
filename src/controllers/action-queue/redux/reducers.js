// @flow

import { type Reducer, combineReducers } from 'redux'

import { type Action } from '../../../types/reduxTypes'
import { type ActionQueueMap } from '../types'

export type ActionQueueState = {
  +queue: ActionQueueMap
}

export const actionQueue: Reducer<ActionQueueState, Action> = combineReducers({
  queue(state: ActionQueueMap = {}, action: Action): ActionQueueMap {
    switch (action.type) {
      case 'ACTION_QUEUE/LOAD_QUEUE': {
        return action.data
      }
      case 'ACTION_QUEUE/QUEUE_ITEM': {
        return {
          ...state,
          [action.programId]: action.item
        }
      }
      case 'ACTION_QUEUE/UPDATE_PROGRAM_STATE': {
        const { programId } = action.state
        const { effect } = action.state

        // Remove program from action queue if it has finished
        if (effect?.type === 'done') {
          const { [programId]: _removed, ...rest } = state
          return rest
        }

        return {
          ...state,
          [programId]: {
            ...state[programId],
            state: action.state
          }
        }
      }
      default:
        return state
    }
  }
})
