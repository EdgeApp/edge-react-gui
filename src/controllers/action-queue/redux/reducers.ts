import { combineReducers } from 'redux'

import type { Action } from '../../../types/reduxTypes'
import type { ActionQueueMap } from '../types'
import { checkEffectIsDone } from '../util/checkEffectIsDone'

export interface ActionQueueState {
  readonly actionQueueMap: ActionQueueMap
  readonly activeProgramIds: string[]
}

export const actionQueue = combineReducers<ActionQueueState, Action>({
  actionQueueMap(state: ActionQueueMap = {}, action: Action): ActionQueueMap {
    switch (action.type) {
      case 'ACTION_QUEUE/LOAD_QUEUE': {
        return action.data
      }
      case 'ACTION_QUEUE/QUEUE_ITEM': {
        return {
          ...state,
          [action.programId]: action.item
        }
      }
      case 'ACTION_QUEUE/UPDATE_PROGRAM_STATE': {
        const { programId } = action.state

        return {
          ...state,
          [programId]: {
            ...state[programId],
            state: action.state
          }
        }
      }
      case 'LOGOUT': {
        return {}
      }
      default:
        return state
    }
  },
  activeProgramIds(state: string[] = [], action: Action): string[] {
    switch (action.type) {
      case 'ACTION_QUEUE/LOAD_QUEUE': {
        const ids = Object.keys(action.data)
        return ids
      }
      case 'ACTION_QUEUE/QUEUE_ITEM': {
        return [...state, action.programId]
      }
      case 'ACTION_QUEUE/UPDATE_PROGRAM_STATE': {
        const { programId } = action.state
        const { effect } = action.state

        // Remove program from action queue if it has finished
        if (checkEffectIsDone(effect)) {
          return state.filter(id => id !== programId)
        }

        return state
      }
      case 'LOGOUT': {
        return []
      }
      default:
        return state
    }
  }
})
