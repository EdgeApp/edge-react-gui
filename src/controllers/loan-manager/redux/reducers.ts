import { combineReducers, Reducer } from 'redux'

import { Action } from '../../../types/reduxTypes'
import { LoanAccountMap } from '../types'

export type LoanManagerState = {
  readonly loanAccounts: LoanAccountMap
}

export const loanManager: Reducer<LoanManagerState, Action> = combineReducers({
  loanAccounts(state: LoanAccountMap = {}, action: Action): LoanAccountMap {
    switch (action.type) {
      case 'LOAN_MANAGER/SET_LOAN_ACCOUNT': {
        const { loanAccount } = action
        return {
          ...state,
          [loanAccount.id]: loanAccount
        }
      }
      case 'LOAN_MANAGER/DELETE_LOAN_ACCOUNT': {
        const { id } = action
        const { [id]: _, ...remainingState } = state
        return remainingState
      }
      case 'LOGOUT': {
        return {}
      }
      default:
        return state
    }
  }
})
