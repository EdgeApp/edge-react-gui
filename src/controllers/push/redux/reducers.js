// @flow

import { type Reducer, combineReducers } from 'redux'

import { type Action } from '../../../types/reduxTypes'
import { type PushEvent } from '../types'

export type PushState = {
  +pushEvents: PushEvent[]
}

export const actionQueue: Reducer<PushState, Action> = combineReducers({
  pushEvents(state: PushEvent[] = [], action: Action): PushEvent[] {
    switch (action.type) {
      case 'PUSH/ADD_PUSH_EVENTS': {
        return [...state, ...action.events]
      }
      default:
        return state
    }
  }
})
