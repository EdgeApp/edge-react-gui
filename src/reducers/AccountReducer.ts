import { combineReducers, Reducer } from 'redux'

import { Action } from '../types/reduxTypes'
import { AccountReferral, Promotion, ReferralCache } from '../types/ReferralTypes'

/**
 * App state tied to the core account object.
 */
export type AccountState = {
  readonly accountReferral: AccountReferral
  readonly accountReferralLoaded: boolean
  readonly referralCache: ReferralCache
  // TODO: Move account settings in here
}

const defaultAccountReferral: AccountReferral = {
  promotions: [],
  ignoreAccountSwap: false,
  hiddenAccountMessages: {}
}

const defaultReferralCache: ReferralCache = {
  accountMessages: [],
  accountPlugins: []
}

const accountInner: Reducer<AccountState, Action> = combineReducers({
  accountReferral(state: AccountReferral = defaultAccountReferral, action: Action): AccountReferral {
    switch (action.type) {
      case 'ACCOUNT_REFERRAL_LOADED': {
        const { referral } = action.data
        // Activating a promo link can race with the intial load,
        // so keep any promotions we already have:
        const promotions = mergePromotions(referral.promotions, state.promotions)
        return { ...referral, promotions }
      }
      case 'ACCOUNT_SWAP_IGNORED': {
        return { ...state, ignoreAccountSwap: action.data }
      }
      case 'MESSAGE_TWEAK_HIDDEN': {
        const { messageId, source } = action.data
        if (source.type === 'promotion') {
          const promotions = state.promotions.map(promo => {
            if (promo.installerId !== source.installerId) return promo
            const hiddenMessages = { ...promo.hiddenMessages, [messageId]: true }
            return { ...promo, hiddenMessages }
          })
          return { ...state, promotions }
        } else {
          const hiddenAccountMessages = { ...state.hiddenAccountMessages, [messageId]: true }
          return { ...state, hiddenAccountMessages }
        }
      }
      case 'PROMOTION_ADDED': {
        const newPromo = action.data
        const promotions = mergePromotions(state.promotions, [newPromo])
        return { ...state, promotions }
      }
      case 'PROMOTION_REMOVED': {
        const installerId = action.data
        const promotions = state.promotions.filter(promo => promo.installerId !== installerId)
        return { ...state, promotions }
      }
    }
    return state
  },

  accountReferralLoaded(state: boolean = false, action: Action): boolean {
    return action.type === 'ACCOUNT_REFERRAL_LOADED' ? true : state
  },

  referralCache(state: ReferralCache = defaultReferralCache, action: Action): ReferralCache {
    switch (action.type) {
      case 'ACCOUNT_REFERRAL_LOADED': {
        const { cache } = action.data
        return cache
      }
      case 'ACCOUNT_TWEAKS_REFRESHED': {
        const cache = action.data
        return cache
      }
    }
    return state
  }
})

// Shared logout logic:
export const account: Reducer<AccountState, Action> = (state: AccountState | undefined, action: Action) => {
  if (action.type === 'LOGOUT') {
    return accountInner(undefined, { type: 'DUMMY_ACTION_PLEASE_IGNORE' })
  }
  return accountInner(state, action)
}

/**
 * Merges lists of promotions, preferring items from b if there's a conflict.
 */
function mergePromotions(a: Promotion[], b: Promotion[]): Promotion[] {
  const toRemove: { [id: string]: true } = {}
  for (const promo of b) toRemove[promo.installerId] = true
  return [...a.filter(promo => !toRemove[promo.installerId]), ...b]
}
