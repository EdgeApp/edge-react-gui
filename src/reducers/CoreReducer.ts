import { Disklet } from 'disklet'
import { EdgeAccount, EdgeContext } from 'edge-core-js/types'
import { combineReducers } from 'redux'

import { Action } from '../types/reduxTypes'

export interface CoreState {
  readonly account: EdgeAccount
  readonly context: EdgeContext
  readonly disklet: Disklet
  readonly otpErrorShown: boolean
}

const flowHack: any = {}
const defaultContext: EdgeContext = flowHack
const defaultDisklet: Disklet = flowHack

const accountHack: any = {
  activeWalletIds: [],
  currencyConfig: {},
  currencyWallets: {},
  watch: () => () => {}
}
export const defaultAccount: EdgeAccount = accountHack

export const core = combineReducers<CoreState, Action>({
  account(state: EdgeAccount = defaultAccount, action: Action): EdgeAccount {
    switch (action.type) {
      case 'LOGIN':
        return action.data.account
      case 'LOGOUT':
        return defaultAccount
      default:
        return state
    }
  },

  context(state: EdgeContext = defaultContext, action: Action): EdgeContext {
    return action.type === 'CORE/CONTEXT/ADD_CONTEXT' ? action.data.context : state
  },

  disklet(state: Disklet = defaultDisklet, action: Action): Disklet {
    return action.type === 'CORE/CONTEXT/ADD_CONTEXT' ? action.data.disklet : state
  },

  otpErrorShown(state: boolean = false, action: Action): boolean {
    switch (action.type) {
      case 'OTP_ERROR_SHOWN':
        return true
      case 'LOGOUT':
        return false
      default:
        return state
    }
  }
})
