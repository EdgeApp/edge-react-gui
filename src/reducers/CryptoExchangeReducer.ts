import { Reducer } from 'redux'

import { Action } from '../types/reduxTypes'
import { GuiCurrencyInfo } from '../types/types'

export type CryptoExchangeState = {
  fromWalletId: string | null
  fromCurrencyCode: string | null
  fromNativeAmount: string
  fromDisplayAmount: string
  fromWalletPrimaryInfo: GuiCurrencyInfo // EdgeCurrencyInfo | null,
  fromBalanceMessage: string

  toWalletId: string | null
  toCurrencyCode: string | null
  toNativeAmount: string
  toDisplayAmount: string
  toWalletPrimaryInfo: GuiCurrencyInfo // EdgeCurrencyInfo | null,
  toBalanceMessage: string

  // Errors:
  insufficientError: boolean
  genericShapeShiftError: string | null

  // Activity flags:
  forceUpdateGuiCounter: number
  shiftPendingTransaction: boolean
  calculatingMax: boolean
}

const dummyCurrencyInfo: GuiCurrencyInfo = {
  walletId: '',
  displayCurrencyCode: '',
  exchangeCurrencyCode: '',
  displayDenomination: {
    name: '',
    multiplier: '1'
  },
  exchangeDenomination: {
    name: '',
    multiplier: '1'
  }
}

const initialState: CryptoExchangeState = {
  fromWalletId: null,
  fromCurrencyCode: null,
  fromNativeAmount: '0',
  fromDisplayAmount: '0',
  fromWalletPrimaryInfo: dummyCurrencyInfo,
  fromBalanceMessage: '',

  toWalletId: null,
  toCurrencyCode: null,
  toNativeAmount: '0',
  toDisplayAmount: '0',
  toWalletPrimaryInfo: dummyCurrencyInfo,
  toBalanceMessage: '',

  insufficientError: false,
  genericShapeShiftError: null,
  forceUpdateGuiCounter: 0,
  shiftPendingTransaction: false,
  calculatingMax: false
}

function cryptoExchangeInner(state = initialState, action: Action): CryptoExchangeState {
  let forceUpdateGuiCounter
  switch (action.type) {
    case 'SELECT_FROM_WALLET_CRYPTO_EXCHANGE': {
      return {
        ...state,
        fromWalletId: action.data.walletId,
        fromWalletPrimaryInfo: action.data.primaryInfo,
        fromCurrencyCode: action.data.currencyCode,
        fromBalanceMessage: action.data.balanceMessage,
        fromNativeAmount: '0',
        toNativeAmount: '0',
        fromDisplayAmount: '0',
        toDisplayAmount: '0',
        // @ts-expect-error
        minerFee: '0',
        genericShapeShiftError: null,
        insufficientError: false
      }
    }

    case 'SELECT_TO_WALLET_CRYPTO_EXCHANGE': {
      return {
        ...state,
        toWalletId: action.data.walletId,
        toCurrencyCode: action.data.currencyCode,
        toWalletPrimaryInfo: action.data.primaryInfo,
        toBalanceMessage: action.data.balanceMessage,
        fromNativeAmount: '0',
        toNativeAmount: '0',
        fromDisplayAmount: '0',
        toDisplayAmount: '0',
        // @ts-expect-error
        minerFee: '0',
        genericShapeShiftError: null
      }
    }

    case 'UPDATE_SWAP_QUOTE': {
      return {
        ...state,
        toNativeAmount: action.data.quote.toNativeAmount,
        toDisplayAmount: action.data.toDisplayAmount,
        fromNativeAmount: action.data.quote.fromNativeAmount,
        fromDisplayAmount: action.data.fromDisplayAmount,
        insufficientError: false,
        genericShapeShiftError: null
      }
    }

    case 'RECEIVED_INSUFFICIENT_FUNDS_ERROR': {
      return {
        ...state,
        insufficientError: true,
        genericShapeShiftError: null
      }
    }

    case 'GENERIC_SHAPE_SHIFT_ERROR': {
      return {
        ...state,
        genericShapeShiftError: action.data
      }
    }

    case 'START_SHIFT_TRANSACTION': {
      return {
        ...state,
        shiftPendingTransaction: true
      }
    }

    case 'DONE_SHIFT_TRANSACTION': {
      return {
        ...state,
        shiftPendingTransaction: false
      }
    }

    case 'START_CALC_MAX': {
      return {
        ...state,
        calculatingMax: true
      }
    }

    case 'SET_FROM_WALLET_MAX': {
      forceUpdateGuiCounter = state.forceUpdateGuiCounter
      forceUpdateGuiCounter++
      return {
        ...state,
        fromNativeAmount: action.data,
        calculatingMax: false,
        forceUpdateGuiCounter
      }
    }

    case 'SHIFT_COMPLETE': {
      return { ...initialState }
    }

    default:
      return state
  }
}

// Nuke the state on logout:
export const cryptoExchange: Reducer<CryptoExchangeState, Action> = (state, action: Action) => {
  if (action.type === 'LOGOUT') {
    return cryptoExchangeInner(undefined, { type: 'DUMMY_ACTION_PLEASE_IGNORE' })
  }

  return cryptoExchangeInner(state, action)
}
