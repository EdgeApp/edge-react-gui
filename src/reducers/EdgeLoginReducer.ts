import { EdgeLobby } from 'edge-core-js'
import { Reducer } from 'redux'

import { Action } from '../types/reduxTypes'

export type EdgeLoginState = {
  lobby: EdgeLobby | null
  error: string | null
  isProcessing: boolean
}

const initialState = {
  lobby: null,
  error: null,
  isProcessing: false
}

export const edgeLogin: Reducer<EdgeLoginState, Action> = (state = initialState, action: Action) => {
  switch (action.type) {
    case 'PROCESS_EDGE_LOGIN': {
      return {
        ...state,
        isProcessing: true
      }
    }

    case 'EDGE_LOBBY_ACCEPT_FAILED': {
      return {
        ...state,
        isProcessing: false
      }
    }

    case 'INVALIDATE_EDGE_LOBBY': {
      return {
        ...state,
        lobby: null,
        isProcessing: false
      }
    }

    case 'SET_LOBBY_ERROR': {
      return {
        ...state,
        lobby: null,
        error: action.data,
        isProcessing: false
      }
    }

    case 'SAVE_EDGE_LOBBY': {
      return {
        ...state,
        lobby: action.data,
        error: null,
        isProcessing: false
      }
    }

    default:
      return state
  }
}
