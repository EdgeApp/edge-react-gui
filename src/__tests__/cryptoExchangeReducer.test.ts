/* globals test expect */

import { cryptoExchange } from '../reducers/CryptoExchangeReducer'

const dummyAction = { type: 'DUMMY_ACTION_PLEASE_IGNORE' }

test('initialState', () => {
  const actual = cryptoExchange(undefined, dummyAction)

  expect(actual).toMatchSnapshot()
})
