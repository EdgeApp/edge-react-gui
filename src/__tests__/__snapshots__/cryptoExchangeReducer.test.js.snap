// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initialState 1`] = `
Object {
  "calculatingMax": false,
  "forceUpdateGuiCounter": 0,
  "fromBalanceMessage": "",
  "fromCurrencyCode": null,
  "fromCurrencyIcon": null,
  "fromCurrencyIconDark": null,
  "fromDisplayAmount": "0",
  "fromNativeAmount": "0",
  "fromWalletId": null,
  "fromWalletPrimaryInfo": Object {
    "displayCurrencyCode": "",
    "displayDenomination": Object {
      "multiplier": "1",
      "name": "",
    },
    "exchangeCurrencyCode": "",
    "exchangeDenomination": Object {
      "multiplier": "1",
      "name": "",
    },
  },
  "genericShapeShiftError": null,
  "insufficientError": false,
  "shiftPendingTransaction": false,
  "toBalanceMessage": "",
  "toCurrencyCode": null,
  "toCurrencyIcon": null,
  "toCurrencyIconDark": null,
  "toDisplayAmount": "0",
  "toNativeAmount": "0",
  "toWalletId": null,
  "toWalletPrimaryInfo": Object {
    "displayCurrencyCode": "",
    "displayDenomination": Object {
      "multiplier": "1",
      "name": "",
    },
    "exchangeCurrencyCode": "",
    "exchangeDenomination": Object {
      "multiplier": "1",
      "name": "",
    },
  },
}
`;
